assets/js/components/UserMenu.js

```diff
@@ -0,0 +1,176 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { Fragment, useState, useRef, useEffect, useCallback } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { clearWebStorage } from '../util';
+import Dialog from './dialog';
+import Button from './button';
+import Menu from './menu';
+import Modal from './Modal';
+import { STORE_NAME as CORE_SITE } from '../googlesitekit/datastore/site/constants';
+import { STORE_NAME as CORE_USER } from '../googlesitekit/datastore/user/constants';
+
+const { useSelect } = Data;
+
+function UserMenu() {
+	const proxyPermissionsURL = useSelect( ( select ) => select( CORE_SITE ).getProxyPermissionsURL() );
+	const userEmail = useSelect( ( select ) => select( CORE_USER ).getEmail() );
+	const userPicture = useSelect( ( select ) => select( CORE_USER ).getPicture() );
+	const splashScreenURL = useSelect( ( select ) => select( CORE_SITE ).getAdminURL( 'googlesitekit-splash', { googlesitekit_context: 'revoked' } ) );
+
+	const [ dialogActive, toggleDialog ] = useState( false );
+	const [ menuOpen, toggleMenu ] = useState( false );
+	const menuButtonRef = useRef();
+	const menuRef = useRef();
+
+	useEffect( () => {
+		global.addEventListener( 'mouseup', handleMenuClose );
+		global.addEventListener( 'keyup', handleMenuClose );
+		global.addEventListener( 'keyup', handleDialogClose );
+
+		return () => {
+			global.removeEventListener( 'mouseup', handleMenuClose );
+			global.removeEventListener( 'keyup', handleMenuClose );
+			global.removeEventListener( 'keyup', handleDialogClose );
+		};
+	}, [] );
+
+	const handleMenu = useCallback( () => {
+		toggleMenu( ! menuOpen );
+	}, [ menuOpen ] );
+
+	const handleMenuClose = useCallback( ( e ) => {
+		if (
+			( ( 'keyup' === e.type && 27 === e.keyCode ) || 'mouseup' === e.type ) &&
+			! menuButtonRef.current.buttonRef.current.contains( e.target ) &&
+			! menuRef.current.menuRef.current.contains( e.target )
+		) {
+			toggleMenu( false );
+		}
+	}, [] );
+
+	const handleDialog = useCallback( () => {
+		toggleDialog( ! dialogActive );
+		toggleMenu( false );
+	}, [ dialogActive ] );
+
+	const handleMenuItemSelect = useCallback( ( index, e ) => {
+		if (
+			( 'keydown' === e.type && ( 13 === e.keyCode || 32 === e.keyCode ) ) || // Enter or Space is pressed.
+			'click' === e.type // Mouse is clicked
+		) {
+			switch ( index ) {
+				case 0:
+					handleDialog();
+					break;
+				case 1:
+					if ( proxyPermissionsURL ) {
+						global.location.assign( proxyPermissionsURL );
+					}
+					break;
+				default:
+					handleMenu();
+			}
+		}
+	}, [ proxyPermissionsURL, handleMenu, handleDialog ] );
+
+	const handleDialogClose = useCallback( ( e ) => {
```

Comment:
Please move the definition of this function within the `useEffect` above where it is bound to the event listeners as it isn't used anywhere else and this avoids potential bugs related to dependencies.